[4 - 6단계 방탈출 사용자 예약] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  https://prolog.techcourse.co.kr/studylogs/4247

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- 스프링 학습에 초점을 맞추었기 때문에 객체지향 생활체조 요구사항을 완벽하게 충족하지는 못했습니다. 그렇지만 최대한 의식적으로 지키려고 노력했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
안녕하세요 로키!
듀이입니다.
빠르게 미션 제출하고 리뷰받고 싶었는데, 이번 주에 테코톡이라는 큰 산을 넘다보니 주말까지 끌고 오게 되었네요🥹
그럼, 이번 리뷰도 시간내어 봐주셔서 감사합니다.
잘 부탁드립니다!🫡

### `Member` 도입으로 인한 db 구조 변경
- 기존에는 `Reservation` 생성 시 예약자 이름 값을 넣어주었는데, 관리자 페이지에서 예약 시 멤버를 선택하고, 그 멤버의 아이디를 통해 예약을 진행해야 하는 방식으로 요구사항이 추가되었습니다.
- 요구사항에 맞춰서 멤버 테이블을 생성하고, 멤버 객체를 만들고 나니 `reservation` 테이블에서 `member`의 id를 외래키로 참조하게 되었는데요. 사실 처음에는 `reservation` 테이블의 구조를 변경하지 않고도 `Member`에서 이름 값만 넣어줄 수도 있었을 것 같다고 생각했습니다. 그렇지만 fk로 참조하게 된 이유는 현재 `reservation` 테이블이 `reservation_time`, `theme` 테이블의 pk를 모두 외래키로 참조하고 있기 때문에 일관성을 보장해주기 위해서 db 구조를 크게 변경하게 되었습니다.

### 패키지 구조
- 현재 패키지 구조가 복잡하지는 않을지 고민입니다. 사실 현재 구조에서 application을 예로 들면 내부에 reservation이나 member와 같은 패키지를 또 추가하고 싶었는데, theme는 reservation에 포함되는 걸까? 그렇다면 member도 reservation에 포함되는 것 아닐까? 이런 생각 때문에 우선은 나누지 않았습니다.
- 로키라면 어떤 방식으로 패키지 구조를 잡아 나가시는지 궁금합니다!

### 통합 테스트
- 서비스 테스트는 최대한 보강하였는데, 통합 테스트는 시간 관계상 많이 보강하지 못했습니다🥲
- `JWT`나 `cookie`, `ArgumentResolver`, `Interceptor` 등과 같이 학습해야 할 키워드들을 우선 경험해보는 것이 더 좋겠다고 생각했기 때문입니다..!
- 부족한 부분이나 제가 잘못 알고 있는 부분이 있는 것 같다면 편하게 말씀해주세요!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

ljhee92 committed:
refactor: Theme 저장 시 파라미터를 개별 값이 아닌 엔티티로 받도록 수정

ljhee92 committed:
refactor: 테스트 시간 고정

ljhee92 committed:
chore: 최근 데이터 업데이트

ljhee92 committed:
feat: 캐치하지 못한 예외들 처리 추가

ljhee92 committed:
refactor: 다건 조회 케이스 size 검증 추가

ljhee92 committed:
refactor: 생성된 결과에서 id를 추출하여 삭제하도록 수정

ljhee92 committed:
refactor: SimpleJdbcInsert 객체 재사용하도록 수정

ljhee92 committed:
refactor: given절의 Reservation을 Fixture로 수정

- ReservationService를 통해 create하지 않고 ReservationDbFixture를 활용하도록 수정

ljhee92 committed:
refactor: 여러 개의 테스트 assertAll로 수정

ljhee92 committed:
docs: 요구사항 명세 추가

ljhee92 committed:
feat: 로그인 폼 페이지 응답

ljhee92 committed:
refactor: infrastructure#repository로 패키지구조 추가, static 변수 위치 수정

ljhee92 committed:
feat: 로그인 기능, 인증 정보 조회 기능 구현

ljhee92 committed:
feat: 클라이언트 코드 수정

ljhee92 committed:
test: 사용자 단위 테스트 추가

ljhee92 committed:
test: 로그인, 인증 조회 서비스 테스트 추가

ljhee92 committed:
refactor: 테스트 DisplayNameGeneration 추가

ljhee92 committed:
refactor: 테스트 클래스 접근제어자 제거

ljhee92 committed:
refactor: 통합 테스트 리팩터링

ljhee92 committed:
refactor: HandlerMethodArgumentResolver를 활용해 컨트롤러의 메서드로 LoginMember를 주입받도록 변경

ljhee92 committed:
refactor: 클라이언트 코드에서 완료한 TODO 제거

ljhee92 committed:
feat: 사용자 예약 생성 기능 변경 - 로그인한 사용자 정보를 활용하도록 리팩터링

ljhee92 committed:
docs: 예약 생성 시 관리자와 사용자 요청 분리

ljhee92 committed:
feat: 클라이언트 코드 수정

ljhee92 committed:
chore: db 구조 변경 (reservation에 member fk 추가)

ljhee92 committed:
refactor: dto 라인 통일

ljhee92 committed:
remove: ReserverName 제거

ljhee92 committed:
feat: 예약 생성 시 Member 주입

ljhee92 committed:
feat: 관리자 예약 생성 요청 분리

ljhee92 committed:
fix: 요청 시 LoginMember와 응답 시 MemberResponse 분리

ljhee92 committed:
test: 통합 테스트 리팩터링

ljhee92 committed:
refactor: 에러 응답 JSON 통일

ljhee92 committed:
refactor: 클라이언트 코드 완료 TODO 제거

ljhee92 committed:
docs: 요구사항 명세 추가

ljhee92 committed:
feat: 쿠키 유효 시간 설정

ljhee92 committed:
chore: db 구조 변경 (member에 role 추가)

ljhee92 committed:
feat: Member에 Role로 권한 추가

ljhee92 committed:
feat: 접근 권한 제어 기능 구현

ljhee92 committed:
test: 테스트 리팩터링

ljhee92 committed:
chore: init 데이터 날짜 수정

ljhee92 committed:
feat: 클라이언트 코드 수정

ljhee92 committed:
feat: 예약 목록 검색 기능 구현

ljhee92 committed:
refactor: 파라미터 final 키워드 통일

ljhee92 committed:
feat: 로그아웃 기능 구현

ljhee92 committed:
test: 서비스 테스트 보강

ljhee92 committed:
test: 관리자 예약 생성 통합 테스트 추가

ljhee92 committed:
refactor: 메서드명 수정

ljhee92 committed:
refactor: request dto 검증 메시지 및 valid 어노테이션 추가

ljhee92 committed:
refactor: id를 포함하는 도메인엔티티 생성 시 정적 팩토리 메서드로 제한하도록 통일

ljhee92 commented on src/test/java/roomescape/domain/MemberTest.java:
```diff
@@ -0,0 +1,27 @@
+
+        assertThat(member.isIncorrectPassword(password)).isFalse();
+    }
+}
```
@coderabbitai 73c853eb 반영했어용

ljhee92 commented on src/main/java/roomescape/application/AuthService.java:
```diff
@@ -0,0 +1,33 @@
+    public AuthService(MemberService memberService, JwtTokenProvider jwtTokenProvider) {
+        this.memberService = memberService;
+        this.jwtTokenProvider = jwtTokenProvider;
+    }
+
+    public String login(LoginRequest loginRequest) {
+        Member member = memberService.findMemberByEmail(loginRequest.email());
+        if (member.isIncorrectPassword(loginRequest.password())) {
+            throw new IllegalArgumentException("[ERROR] 올바르지 않은 비밀번호입니다.");
+        }
+        return jwtTokenProvider.createToken(loginRequest.email());
+    }
```
@coderabbitai 사용자에게 편의성을 위해 이메일이나 비밀번호 중 어느 것이 올바르지 않았는지 알려주는 것이 더 좋지 않을까요?

Rok93 replied:
공격할 여지를 주지않는다는 면도 납득은 가지만, UX 친화적이지는 못할 것 같다는 듀이의 의견에도 공감합니다. 😃
서비스의 성격에 따라서 트레이드 오프를 따지며 선택을 해야할 부분이라고 생각되네요.

ljhee92 replied:
저도 토끼의 리뷰를 받고 더 찾아봤어요!
보안이 중요한 서비스(대체로 은행)는 어느 정보가 틀렸는지 알려주지 않고 한 번에 검증하되 내부적으로 로그는 찍어두고,
사용자 경험이 중요한 서비스(쇼핑이나 배달)는 사용자 친화적으로 구현한다고 하더라구요!
저는 개인적으로 사용자일 때 아이디와 비밀번호 중 어느 것이 틀렸는지 알려주지 않았을 때 불편함을 더 많이 느꼈어서 사용자 경험을 더 중시하는 편인데 서비스의 성격에 따라 보안의 중요성도 한 번 더 깨달았습니다😄

ljhee92 committed:
refactor: 토큰 유효성 검증 추가

ljhee92 commented on src/main/java/roomescape/application/AuthService.java:
```diff
@@ -0,0 +1,33 @@
+        String payload = jwtTokenProvider.getPayload(token);
+        Member member = memberService.findMemberByEmail(payload);
+        return new LoginMember(member.getId(), member.getName(), member.getRole(), member.getEmail());
+    }
```
@coderabbitai c72bd281 반영했어용

ljhee92 committed:
refactor: 토큰 추출 중복 코드 분리

ljhee92 commented on src/main/java/roomescape/presentation/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,50 @@
+                .findFirst()
+                .map(Cookie::getValue)
+                .orElseThrow(() -> new AuthException("[ERROR] 로그인이 필요합니다.", HttpStatus.UNAUTHORIZED));
+    }
```
@coderabbitai 08e7f403 반영했어용

ljhee92 commented on src/main/java/roomescape/presentation/AuthenticationPrincipalResolver.java:
```diff
@@ -0,0 +1,53 @@
+                .findFirst()
+                .map(Cookie::getValue)
+                .orElseThrow(() -> new AuthException("[ERROR] 토큰이 존재하지 않습니다.", HttpStatus.UNAUTHORIZED));
+    }
```
@coderabbitai 08e7f403 반영했어용

ljhee92 committed:
fix: Auth 예외 코드 수정

ljhee92 commented on src/main/java/roomescape/presentation/exception/GlobalExceptionHandler.java:
```diff
@@ -3,38 +3,69 @@
+        ErrorResponse errorResponse = ErrorResponse.of(HttpStatus.CONFLICT, message);
+        return ResponseEntity
+                .status(e.getStatus())
+                .body(errorResponse);
```
@coderabbitai 헉 복붙의 폐해..! f07e9131 반영했어용

ljhee92 committed:
fix: 메서드명 수정

ljhee92 commented on src/main/java/roomescape/application/ReservationService.java:
```diff
@@ -64,4 +94,15 @@ private Reservation findReservationById(Long id) {
+            LocalDate dateTo
+    ) {
+        List<Reservation> reservations = reservationRepository.findAllByMemberIdAndThemeIdAndDateBetween(themeId, memberId, dateFrom, dateTo);
+
```
@coderabbitai 헉 이건 메서드명을 수정해아해용 591af69d 반영했어용

ljhee92 committed:
fix: 예악시간 삭제 시 예약 아이디를 전달하고 있었던 것 수정

ljhee92 commented on src/test/java/roomescape/application/ReservationTimeServiceTest.java:
```diff
@@ -26,62 +37,99 @@ public class ReservationTimeServiceTest extends BaseTest {
 
-        assertThatThrownBy(() -> reservationTimeService.deleteReservationTimeById(1L))
+        assertThatThrownBy(() -> reservationTimeService.deleteReservationTimeById(reservation.getId()))
                 .isInstanceOf(IllegalArgumentException.class);
```
@coderabbitai 꼼꼼하구만.. c3541224 반영했어용

ljhee92 committed:
refactor: 오타 수정

ljhee92 commented on src/test/java/roomescape/integration/ReservationTest.java:
```diff
@@ -1,220 +1,463 @@
+            jdbcTemplate.update("INSERT INTO theme (name, description, thumbnail) VALUES (?, ?, ?)",
+                    "테마1", "설명1", "썸네일1");
+            jdbcTemplate.update("INSERT INTO member (name, role, email, password) VALUES (?, ?, ?, ?)",
+                    "브라운", "USER", "test@eamil.com", "pass1");
```
@coderabbitai ㅠㅠ c894d5d6 반영했어용

Rok93 commented on src/main/java/roomescape/application/ReservationService.java:
```diff
@@ -37,24 +44,47 @@ public List<ReservationResponse> getReservations() {
     }
 
-    public ReservationResponse createReservation(ReservationCreateRequest request) {
+    public ReservationResponse createReservation(ReservationCreateRequest request, LoginMember loginMember) {
```
> 요구사항에 맞춰서 멤버 테이블을 생성하고, 멤버 객체를 만들고 나니 reservation 테이블에서 member의 id를 외래키로 참조하게 되었는데요. 사실 처음에는 reservation 테이블의 구조를 변경하지 않고도 Member에서 이름 값만 넣어줄 수도 있었을 것 같다고 생각했습니다. 

동명이인의 케이스를 생각해보면 이방식은 한계점이 있습니다. 
하다못해 email과 같은 unqiue한 속성이라면 모를까, 그 외의 중복으로 존재할 수 있는 데이터의 경우 예약 속성으로 넣으면 추후 어떤 유저의 예약인지 찾을 수 없는 문제가 발생할 수 있다고 생각해요. 이 외에도 지금은 unique지만 나중에 해당 속성이 중복을 허용하도록 정책이 변경된다면 또 문제가 생길 수 있을 것 같아요. 🤔

결론은 추적을 하기위해서는 식별자인 id를 가지는 것이 맞다고 생각해요.

ljhee92 replied:
와.. 동명이인 경우의 수는 생각하지 못했네요!
로키의 꼼꼼함에 다시 한 번 감탄했어용👍

사실 저 reservation 테이블 구조를 변경하지 않아도 되지 않을까 라는 생각은 '학습할 때는 외래키를 사용하도록 배우는데 실무에서는(정답이 아니라는 것을 알고 있지만) 왜 외래키를 잘 사용하지 않는다고 할까?'로부터 생겨난 거긴 합니다!
좀 더 찾아보기도 하고 로키의 말씀을 듣고 나니 구분할 수 있는 유니크한 값은 필요하다고 생각이 정립되었어요.
외래키를 사용하지 않는다는 것은 reservation에 식별자를 넣되 db에서 외래키로 강제하지 않는 것이고 reservation_member와 같은 중간 테이블을 두는 것 같더라고요! (저도 구글링과 지피티에게 물어나온 내용이니 아니라면 정정 부탁드립니다!)

Rok93 replied:
`실무에서 외래키를 사용하지 않는 방식`을 선호한다는 말은 제 사례 혹은 주변의 이야기만 들어봤을 때는 실제로 그렇게하는 경우가 많은 걸로 알고 있어요. 😃 

생각나는대로 이유를 나열해보자면...

* 애플리케이션 레벨에서 무결성을 관리해도 충분히 무결성을 관리할 수 있음 
* 운영 환경에서의 유연성 부족 
  * 데이터 삭제/수정 시 제약 조건 위반이 발생해서 제약이 생길 수 있고 대규모 배치 작업이나 DB  마이그레이션이 어려워질 수 있음
* 성능상의 이슈
  *  FK가 있으면 CUD 시마다 제약 조건을 검증해야 하므로 성능 오버헤드가 발생.

누락된 내용이 있을지 모르겠는데, 제가 알기로는 이런 이유 때문에 FK를 안쓰는 방식을 많이 쓰고있는 것으로 알고있어요.

Rok93 commented on src/main/java/roomescape/application/ReservationService.java:
```diff
@@ -37,24 +44,47 @@ public List<ReservationResponse> getReservations() {
+
+    private ReservationDateTime getReservationDateTime(Long timeId, ReservationDate reservationDate) {
+        ReservationTime reservationTime = reservationTimeService.findReservationTimeById(timeId);
+        return ReservationDateTime.create(reservationDate, reservationTime, currentTimeService.now());
```
예약 시간이 유효하지 않은 경우, 현재 500 응답을 내려주고있는데요. 
듀이는 400대 응답과 500대 응답은 각각 어떤 상황에서 내려주고계신가요? 😃

<img width="1815" alt="image" src="https://github.com/user-attachments/assets/f33f4e4b-1eb7-4380-9209-413b3f7217f6" />

ljhee92 replied:
헉 예약 시간이 유효하지 않은 경우가 어떤 경우인지 한 번 더 상세하게 말씀해주실 수 있을까요?
공유해주신 사진으로만 유추해보자면 현재시간 이전의 예약 시간을 선택했을 때라고 생각되는데요..!
저도 같은 상황에서 실행해보았는데, 400 응답을 내려주고 있는 것으로 확인되었어요!

<img width="1706" alt="스크린샷 2025-05-13 오후 9 16 02" src="https://github.com/user-attachments/assets/d631e68b-4d65-42a0-9fcd-592638805564" />


질문주신 물음에 답해보자면, 400대 응답은 클라이언트 측에서 발생한 에러에 사용하고 있습니다.
예를 들어 잘못된 입력이 들어왔거나 인증되지 않은 사용자일 경우 혹은 권한이 없는 사용자일 경우 등이요!

500대 응답은 서버 측에서 발생한 에러일 경우 사용하고 있습니다.
다른 코멘트에서 말씀해주신 것처럼 클라이언트에게 상세 예외 메시지를 내려주지 않는 것도 서버에서만 알고 있어도 되는 에러이기 때문이라고 생각합니다.

Rok93 replied:
엇.. 저는 저 상황에서 500 응답이 내려왔었거든요. 🤔
500 응답은 서버측 에러고 보통 의도하지 않은 경우에 내려주는 경우이다보니 지금과 같이 유효성검증에서 걸려서 내려준 응답에서 500대 응답은 적절하지 않다는 의도였는데, 제가 리뷰하면서 코드를 조금 건드리기도하다보니 작성하신 코드와 다르게 동작한건가....? 의심이되네요. 😅 

400 응답이 내려오고있다면 이 부분에 대한 코멘트는 그냥 넘어가셔도 될 것 같습니다. 🙏

Rok93 commented on src/main/java/roomescape/presentation/exception/ReservationExceptionHandler.java:
```diff
@@ -5,19 +5,11 @@
-        return ResponseEntity.badRequest().body(errorResponse);
-    }
-
     @ExceptionHandler(PastReservationException.class)
```
ReservationExceptionHandler와 GlobalExceptionHandler 둘 다 같은 위치에있다면 따로 분리한 의미가 없는 것 같아요. 👀
둘을 차라리 하나로 합치는건 어떨까요? 😃

ljhee92 replied:
맞아요 그렇네요!!
로키 의견에 동의합니다! 굳이 나눌 필요 없이 `GlobalExceptionHandler`에서 한 번에 처리할 수 있도록 변경해주었습니다.
f82cd4b4 에 반영했습니다~!

Rok93 commented on src/main/resources/data.sql:
```diff
@@ -16,42 +24,59 @@ VALUES ('지하 감옥',
-VALUES ('오래된예약자18', '2025-04-12', 1, 1);
-INSERT INTO reservation (name, date, time_id, theme_id)
-VALUES ('오래된예약자19', '2025-04-11', 2, 2);
+-- reservation: 오래된 데이터
```
더미데이터 꼼꼼하게 넣어주셔서 감사해요. 🙏🙏
덕분에 리뷰할 때, 조금 더 수월하게 진행할 수 있었어요. ☺️

ljhee92 replied:
PR 내용에 확인하실 수 있는 테스트 계정을 표기하는 걸 깜빡해서 데이터 찾기 번거로우셨을텐데도 좋게 말씀해주셔서 감사합니다 로키🥹👍

Rok93 commented on src/main/java/roomescape/presentation/dto/response/MemberResponse.java:
```diff
@@ -0,0 +1,23 @@
+        return new MemberResponse(member.getId(), member.getName(), member.getEmail());
+    }
+
+    public static List<MemberResponse> from(List<Member> members) {
```
사소하지만 `List<MemberResponse>`를 반환하는거라면 `toList`와 같은 네이밍은 어떨까요? 😃

ljhee92 replied:
전혀 사소하지 않습니다! 메서드 호출해서 사용할 때 헷갈릴 수 있는 부분이라서 로키 의견대로 `toList`로 변경해주었습니다😄

e3dee125 반영한 커밋입니다!

Rok93 commented on src/main/java/roomescape/presentation/exception/GlobalExceptionHandler.java:
```diff
@@ -3,38 +3,69 @@
+        String message = PREFIX_FOR_ERROR_MESSAGE + e.getMessage();
+        log.error(message, e);
+        ErrorResponse errorResponse = ErrorResponse.of(HttpStatus.INTERNAL_SERVER_ERROR, message);
+        return ResponseEntity.internalServerError().body(errorResponse);
```
토끼의 말처럼 이 케이스는 우리가 예상하지 못한 에러를 캐치하는 것이기 때문에 에러 메시지를 응답으로 바로 내려주는 것은 조금 우려되는면이있는 것 같아요. 👀

(실제 시스템이라면...) log로 에러메시지를 확인할 수 있을 것이니 에러응답으로는 "알 수 없는 에러가 발생하였습니다." 와 같은 메시지 정도로 내려줘도 괜찮지 않을까싶네요. 🤔

ljhee92 replied:
클라이언트까지 내려줘야 할 예외 메시지도 보안 측면까지 잘 고려해야 한다는 점을 간과하고 있었어요.
덕분에 한 가지 더 배웠습니다!
c1915cb0 에 반영했습니다.

Rok93 commented on src/main/java/roomescape/application/AuthService.java:
```diff
@@ -0,0 +1,38 @@
+        if (!jwtTokenProvider.validateToken(token)) {
+            throw new AuthException("[ERROR] 유효하지 않은 토큰입니다.", HttpStatus.UNAUTHORIZED);
+        }
+        String payload = jwtTokenProvider.getPayload(token);
```
여기서 얻어온 payload는 email인 것 같은데  변수명을 email이라고 명시적으로 두는건 어떨까요? 😃

ljhee92 replied:
앗 그렇네요!! 저도 처음에 payload가 뭐지? 생각했으면서 명확하게 해줄 생각은 못했습니다.
655ca019 에 반영해주었습니다!
추가로, `AuthService#login`에서 `JwtTokenProvider#createToken` 메서드를 호출할 때 `loginRequest`의 email을 인자로 넘겨주고 있었는데, 패스워드 검증을 완료한 `member`의 email을 넘겨주는 것으로 함께 수정했습니다.

Rok93 commented on src/main/java/roomescape/infrastructure/jwt/JwtTokenProvider.java:
```diff
@@ -0,0 +1,63 @@
+package roomescape.infrastructure.jwt;
```
jwt도 infrastructure 패키지 하위에있군요? 🤔
듀이가 생각하는 infrastructure는 어떤 계층인가요?

ljhee92 replied:
저는 infrastructure 계층을 외부 의존성과 연결되어 있는 계층이라고 생각하고 있습니다.
현재는 dao에 가깝긴 하지만 repository라고 명명한 것들이 infrastructure 계층에 위치하고 있는 이유도 db라는 외부 의존성과 연결되어 있기 때문이고요!
같은 이유로 jwt도 jjwt 라이브러리라는 외부 의존성과 연결되어 있기 때문에 infrastructure 계층에 위치시켰습니다.
이런 의도를 담아 위치시킨 것이었는데 물어봐주셔서 너무 좋네용!
로키의 의견은 어떠신지도 궁금합니다!!😆

Rok93 replied:
`모든 외부 라이브러리 사용이 곧 infrastructure` 라는 극단적인 생각만 아니라면 좋은 의도라고 생각해요. 😃

ljhee92 replied:
오 네!
안 그래도 로키에게 제 의도를 설명드릴 때 Lombok이나 Validation 같은 경우는
외부 라이브러리를 사용하지만 infrastructure 계층에 위치하지 않는다고 생각했었어요..!
모든 외부 라이브러리 사용이 곧 infrastructure 계층이라는 극단적인 생각을 하지는 않도록 주의하겠습니다👍🏻

Rok93 commented on README.md:
```diff
@@ -1,6 +1,6 @@
 # 방탈출 사용자 예약
 
-## 1단계
```
> 현재 패키지 구조가 복잡하지는 않을지 고민입니다. 사실 현재 구조에서 application을 예로 들면 내부에 reservation이나 member와 같은 패키지를 또 추가하고 싶었는데, theme는 reservation에 포함되는 걸까? 그렇다면 member도 reservation에 포함되는 것 아닐까? 이런 생각 때문에 우선은 나누지 않았습니다.

service 계층은 유즈케이스 단위로 나누는 것이 일반적이라 생각해요. 😃
예를들어 `theme는 reservation에 포함되는 걸까?` 와 같은 경우 reservation을 저장하는 경우에는 포함될 수 있지만 신규 theme를 추가하는 경우라거나 단순히 theme를 조회하는 경우 등 theme를 CRUD하는 경우에는 reservation과는 상관없이 순수하게 theme만의 service 클래스가 필요하다고 생각해요. 🤔 

지금의 경우에는 계층형 아키텍처를 선택한 것으로 보여요. 
이런 경우라면 오히려 각 계층 패키지마다 도메인별로 패키지를 쪼개는 것이 더 패키지 구조를 파악하기가 어려워질 것 같아요. 👀 각 도메인에 관련된 클래스 파일이 많다면 저라면 도메인형 패키지 구조를 택할 것 같아요. 

[[아키텍쳐] 패키지 구조 : 계층형 VS 도메인형 어떤 것을 선택할까?](https://ksh-coding.tistory.com/96)

ljhee92 replied:
지난 미션에서 계층형 아키텍쳐를 도입한 이유를 망각하고 단순히 클래스가 조금 증가했다고 이런 생각을 하게 된 것 같네요..! 하핫😅
로키가 주신 아티클도 읽어보고 다시 한 번 계층형과 도메인형 아키텍쳐의 장단점을 생각해봤어요.
- 계층형 아키텍쳐
  - 장점: 프로젝트 구조를 파악하기 쉬움, 도메인이 많지 않은 경우 적합
  - 단점: 변경이 있을 때 미치는 영향의 범위가 넓다. (-> 아마 member가 추가되면서 제가 느꼈던 단점인 것 같아요.)
- 도메인형 아키텍쳐
  - 장점: 도메인별 응집도가 높다.
  - 단점: 프로젝트 흐름을 파악하기 어렵다.

아직까지는 스프링에서의 레이어드 아키텍쳐를 공부하는 것이 우선이라고 생각하기도 하고, 도메인이 크게 많다고 느껴지지 않기 때문에 계층형을 유지하려고 합니다.
추후에 도메인이 더 추가되어야 한다거나 불편함이 느껴진다면 그 때 도메인형 아키텍쳐로의 변경을 고려해봐도 좋을 것 같아요!

Rok93 reviewed (changes requested):
듀이 4~6단계도 잘 수행해주셨어요. 👍💯
전체적으로 잘 구현해주셔서 크게 피드백드릴 부분이 없었네요. 🙄
질문에대한 답변과 몇몇 코멘트 남겨두었는데, 확인 부탁드릴게요. 🙏

ljhee92 committed:
refactor: entity -> dto 변환시 List는 toList로 메서드명 수정

ljhee92 committed:
refactor: token 생성시 매개변수 payload -> email로 명시

ljhee92 committed:
refactor: 500 에러 발생시 예외 메시지를 상세하게 내려주지 않도록 변경

ljhee92 committed:
refactor: 예외 처리 한 곳에서 하도록 변경

ljhee92 committed:
refactor: Role도 JWT 토큰에 추가하도록 변경

- Interceptor에서 Role을 검증할 때 AuthService#findMemberByToken을 통해 Member를 찾을 때 DB까지 찔러줘야 했는데, Token을 통해서만 Role을 검증할 수 있다는 장점이 있어 변경했습니다.

ljhee92 commented:
안녕하세요 로키!
구현하면서는 `ArgumentResolve`와 `Interceptor`가 너무 어렵고 `Member` + `/admin` 도입으로 깨지는 테스트가 많았어서 힘들었는데, 잘 구현했다고 말씀해주시니 덕분에 자신감 충전할 수 있었습니다🥹👍🏻
항상 소중한 시간 들여서 꼼꼼하게 리뷰해주셔서 너무너무 감사드려요!
그럼 이번 리뷰 요청도 잘 부탁드립니다🙇‍♀️

Rok93 commented on src/main/java/roomescape/application/AuthService.java:
```diff
@@ -0,0 +1,47 @@
+        String email = jwtTokenProvider.getEmail(token);
+        Member member = memberService.findMemberByEmail(email);
+        return new LoginMember(member.getId(), member.getName(), member.getRole(), member.getEmail());
+    }
```
고민해볼 부분인데요. 😃
지금처럼 매번 token을 통해서 추출한 email로 member를 조회한다면 (로그인이 필요한) API 요청은 모두 member 테이블 조회가 1번은 무조건 필요할 것 같은데요. 🤔

토큰만있다면 별도의 user 정보 조회없이도 인증을해줄 수 있지 않을까요? 
그렇게된다면 로그인 혹은 권한과 관련된 API를 수행할 때, DB 조회 로직이 하나 줄어들 것 같은데요. 

추가적으로 role 정보도 token에 포함시켜준다면, 권한에 따른 인가도 member 테이블 조회없이 가능할 것 같아요. 😃

이 부분에 대해서는 어떻게 생각하시나요?

ljhee92 replied:
사실 이 부분은 02ad000b 커밋에서 `Role`을 `token`에 포함시키면서 고민하기도 했던 부분인데요.

처음에는 그러려면 `AuthenticationPrincipalResolver`가 `JwtTokenProvider`를 참조해야 한다고 생각했어요.
그렇게 된다면 presentation 계층이 service 계층만 참조하는 것이 아니라 infrastructure 계층까지 참조하게 되는 것이고, 이는 올바른 계층별 참조가 아니라고 판단했구요!
그래서 우선 위 커밋내역처럼 `Role`만 `token`에 넣도록 적용해주었는데 로키가 이렇게 또 저에게 생각할 부분을 주셨네요!🤣👍🏻

`AuthenticationPrincipalResolver`가 `JwtTokenProvider`를 참조하지 않고도 인증할 수 있는 방법을 고민해보니, `token`에 `LoginMember`에 필요한 정보를 함께 담으면 된다고 생각했어요.
`token`에 들어가는 정보는 보안 상 중요한 정보가 아니라면 괜찮다고 생각하기 때문에 이 정보들을 모두 추가해주었습니다.
61e697b2 에 반영해두었습니다!😄

혹시 제가 구현한 방법 외에 로키가 의도하셨던 방법이 다르다면 공유해주실 수 있을까요?~

ljhee92 replied:
로키 안녕하세요! 아직 확인 전이신 듯하여 추가 코멘트 남겨봅니다.
다음 미션을 제 코드 기반으로 작성하다보니 페어와 해당 건에 대해 논의하게 되었는데요!
토큰에 너무 많은 개인 정보를 포함하게 된다면 토큰이 탈취되었을 때 문제가 될 수 있을 것 같습니다.
그래서 만약 한 번 더 리팩터링한다면 토큰과 `LoginMember`에 id, name, role까지만 포함하고, email은 제외할 것 같습니다!

Rok93 replied:
사실 email이 엄청 중요한 정보로 볼 것인가는 조금 애매한면도 있다고 생각하는데 현재는 email이 곧 loginId이기 떄문에 email을 숨기는게 좋을 것 같다고 생각되기도하네요. 🤔

Rok93 commented on src/main/java/roomescape/application/MemberService.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    public MemberResponse createMember(MemberCreateRequest request) {
+        Member member = Member.create(request.name(), Role.USER, request.email(), request.password());
```
사소하지만 이런 경우 직접 Role을 생성자 파라미터로 주입하기보다는 정적 팩토리 메서드로 `ofUser` 와 같이 USER 생성을 위한 정적 팩토리 메서드를 제공해주는 것도 괜찮을 것 같아요. 😃 

지금의 경우에는 이 로직 외에는 Member 생성 로직이 없을 것 같아서 굳이 변경하지는 않아도 될 것 같네요. 😉

ljhee92 replied:
그렇게 한다면 정적 팩토리 메서드의 장점을 더 활용할 수 있을 것 같아요!
기억해두고 있다가 추후에 `ofUser` 혹은 `ofAdmin` 등으로 나뉘어야 할 때 활용해보도록 하겠습니다👍🏻

Rok93 reviewed (changes requested):
듀이 피드백 반영 잘해주셨어요. 👍👍 
몇몇 코멘트와 추가 질문 그리고 추가적으로 제 생각 남겨두었는데 같이 확인 부탁드릴게요. 🙏
지금 당장 머지해도 괜찮을 정도로 충분히 잘 구현해주셨지만...! 
혹시 마지막으로 수정할 부분이 있을까봐 request change 드렸습니다. 
다음 리뷰 요청 때는 바로 머지하도록할께요. 😉

ljhee92 committed:
refactor: LoginMember의 정보 JWT 토큰에 추가하도록 변경

- Role과 마찬가지로 토큰으로 LoginMember를 찾을 때 DB까지 접근하지 않도록 하기 위해 변경했습니다.

ljhee92 commented:
안녕하세요 로키!
리뷰 요청 기간이 지났음에도 꼼꼼하게 마지막까지 리뷰해주셔서 감사드려요!
제게 질문주시는 것에 더해 로키의 의견도 들을 수 있어서 좋았습니다!😄👍🏻

Rok93 commented on src/main/java/roomescape/presentation/api/AuthController.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    @PostMapping("/login")
+    public ResponseEntity<Void> login(@RequestBody LoginRequest loginRequest, HttpServletResponse response) {
```
`HttpServletResponse`를 사용하는 방식 대신에 아래와 같이 ResponseEntity의 header를 통해 cookie를 지정하는 방법은 어떨까요? 

```
    return ResponseEntity.ok()
            .header(HttpHeaders.SET_COOKIE, "token", "tokenValue")
            .build();
```

ljhee92 replied:
좋은 것 같아요!
매 번 `HttpServletResponse`를 매개변수로 받기보다 `ResponseEntity`의 header를 통해 지정해주도록 변경해보았습니다.

6595df6c 에 반영했습니다!

Rok93 reviewed (approved):
듀이 마지막까지 피드백 반영 잘해주셨습니다. 👍👍 
방탈출 사용자 예약 미션은 이만 종료하도록하겠습니다. 👏👏 
고생많으셨어요! 다음 미션도 화이팅입니다. 💪💪💪
